akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -151,8 +151,16 @@ private[http] object HttpServerBluePrint {
         // and then toggle back to "idle"
 
         val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
-        // optimization: re-use the idle outHandler
-        entitySource.setHandler(this)
```

Comment:
This optimization can be restored in a similar way to client's `ResponsePrep` stage by moving entity stream to a variable and implementing `onDownstreamFinish` as a member of stage itself. Things look a bit ..."convoluted" on the server side though, so I decided to avoid messing with it for now (: